tests/bluetooth/host/keys/bt_keys_clear/src/main.c

```diff
@@ -0,0 +1,210 @@
+};
+
+/* Global iterator to iterate over the ID, Address and type LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_pair const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+static struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Clear an existing key and verify the result
+ *
+ *  Constraints:
+ *   - Key reference points to a valid item
+ *
+ *  Expected behaviour:
+ *   - The key content is cleared
+ */
+void test_verify_key_cleared_correctly(void)
+{
+	struct bt_keys empty_key;
+	struct bt_keys *key_ref_to_clear, *find_returned_ref;
+	uint8_t id_to_find = current_params_vector->id;
+	const bt_addr_le_t *addr_to_find = current_params_vector->addr;
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		ztest_test_skip();
+	}
+
+	memset(&empty_key, 0x00, sizeof(struct bt_keys));
+	/* 'params_it' is incremented before running the test case */
+	key_ref_to_clear = returned_keys_refs[params_it - 1];
+
+	/* Ensure that item exists in the keys pool */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
+	zassert_true(find_returned_ref != NULL, "bt_keys_find_addr() returned a NULL reference");
+
+	bt_keys_clear(key_ref_to_clear);
+
+	expect_not_called_bt_id_del();
+
+	/* Verify that memory was cleared */
+	zassert_mem_equal(key_ref_to_clear, &empty_key, sizeof(struct bt_keys),
+			  "Key content wasn't cleared by 'bt_keys_clear()'");
+
+	/* Ensure that item doesn't exist in the keys pool after calling bt_keys_clear() */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
+	zassert_true(find_returned_ref == NULL,
+		     "bt_keys_find_addr() returned a non-NULL reference");
+}
+
+/*
+ *  Clear an existing key and verify the result
+ *
+ *  Constraints:
+ *   - Key reference points to a valid item
+ *
+ *  Expected behaviour:
+ *   - The key content is cleared
+ *   - bt_id_del() is called with correct key reference
+ */
+void test_verify_key_cleared_correctly_and_bt_id_del_called(void)
+{
+	struct bt_keys empty_key;
+	struct bt_keys *key_ref_to_clear, *find_returned_ref;
+	uint8_t id_to_find = current_params_vector->id;
+	const bt_addr_le_t *addr_to_find = current_params_vector->addr;
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		ztest_test_skip();
+	}
+
+	memset(&empty_key, 0x00, sizeof(struct bt_keys));
+	/* 'params_it' is incremented before running the test case */
+	key_ref_to_clear = returned_keys_refs[params_it - 1];
+
+	/* Ensure that item exists in the keys pool */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
+	zassert_true(find_returned_ref != NULL, "bt_keys_find_addr() returned a NULL reference");
+
+	bt_keys_clear(key_ref_to_clear);
+
+	expect_single_call_bt_id_del(key_ref_to_clear);
```

Comment:
I would be concerned that this approach (having one function that tests that a flag results in a behavior and a completely different function that checks that the absence of that flag has a different behavior could result in a combinatoric explosion of cases as you get into more complex modules. If you configure your cases with a mix of parameters and use whether the flag was set before clear to determine whether the mock of bt_id_del was called you could simplify this quite a bit.

Reply:
Yes, your point is totally valid, till now I'm trying to avoid adding a lot of checks to make the test as concise as possible and only testing one path.

Also, they previously complained about adding branching inside the test case, but it's hard to avoid them when the behavior is more dependent on different configuration parameters